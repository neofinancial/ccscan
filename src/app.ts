import fs from 'fs';
import path from 'path';
import chalk from 'chalk';
import globby from 'globby';
import yargs, { Argv } from 'yargs';
import luhn from 'fast-luhn';
import lineReader from 'line-reader';

export interface Args {
  exclude?: string[];
  files: string;
  'luhn-check'?: boolean;
  silent?: boolean;
  verbose?: boolean;
}

const defaultArgs = {
  exclude: [],
  files: '**/*.{ts,js}',
  'luhn-check': true,
  silent: false,
  verbose: false
};

const isCardNumber = (suspect: string): boolean => luhn(suspect.replace(/[\s-]+/g, ''));

const showMatches = async (
  fileName: string,
  matches: RegExpMatchArray,
  args: Args
): Promise<void> => {
  return new Promise((resolve): void => {
    const filePath = path.isAbsolute(fileName) ? fileName : path.join(process.cwd(), fileName);
    const outputLines: string[] = [];
    let lineNumber = 0;
    let warning = false;
    let error = false;

    lineReader.eachLine(filePath, (line: string, last?: boolean): void => {
      lineNumber++;

      for (const match of matches) {
        if (line.match(new RegExp(`.*${match}.*`))) {
          const columnNumber = line.search(match) + 1;

          if (args['luhn-check'] && isCardNumber(match)) {
            error = true;
            outputLines.push(
              `${chalk.gray(`${lineNumber}:${columnNumber}`.padEnd(8))} ${chalk.red(
                'error'.padEnd(6)
              )} ${line.replace(match, chalk.red(match))}`
            );
          } else if (!args['luhn-check']) {
            if (isCardNumber(match)) {
              error = true;
              outputLines.push(
                `${chalk.gray(`${lineNumber}:${columnNumber}`.padEnd(8))} ${chalk.red(
                  'error'.padEnd(6)
                )} ${line.replace(match, chalk.red(match))}`
              );
            } else {
              warning = true;
              outputLines.push(
                `${chalk.gray(`${lineNumber}:${columnNumber}`.padEnd(8))} ${chalk.yellow(
                  'warn'.padEnd(6)
                )} ${line.replace(match, chalk.yellow(match))}`
              );
            }
          }

          break;
        }
      }

      if (last) {
        if ((!args.silent && outputLines.length > 0 && error) || (warning && !args['luhn-check'])) {
          console.log(fileName);
          console.log();
          outputLines.forEach((line: string): void => {
            console.log(`  ${line}`);
          });
          console.log();
        }

        resolve();
      }
    });
  });
};

const scanFile = async (fileName: string, args: Args): Promise<boolean> => {
  const contents = fs.readFileSync(fileName, 'utf-8');
  const matches = args['luhn-check']
    ? contents.match(/\b[2-6]{1}\d{3}[\s-]*\d{4}[\s-]*\d{4}[\s-]*\d{4}[\s-]*\b/gm)
    : contents.match(/\b[1-9]{1}\d{3}[\s-]*\d{4}[\s-]*\d{4}[\s-]*\d{4}[\s-]*\b/gm);

  if (matches && matches.length > 0) {
    !args.silent && (await showMatches(fileName, matches, args));

    return true;
  } else {
    !args.silent && args.verbose && console.log(chalk.gray(fileName));

    return false;
  }
};

const scanFiles = async (args: Args): Promise<boolean> => {
  const mergedArgs = { ...defaultArgs, ...args };
  let found = false;

  try {
    const pattern = [mergedArgs.files, '!node_modules'];

    mergedArgs.exclude &&
      mergedArgs.exclude.forEach((exclude: string): void => {
        pattern.push(`!${exclude}`);
      });

    const files = await globby(pattern);

    for (const fileName of files) {
      if (await scanFile(fileName, mergedArgs)) {
        found = true;
      }
    }

    return found;
  } catch (error) {
    if (!mergedArgs.silent) {
      throw error;
    }

    return found;
  }
};

const run = async (): Promise<void> => {
  yargs
    .scriptName('ccscan')
    .usage('$0 [files] [options]')
    .command({
      command: '$0 [files] [options]',
      describe: 'Scan files for credit card numbers',
      builder: (yargs): Argv<{ files: string }> =>
        yargs.positional('files', {
          default: '**/*.{ts,js}',
          describe: 'Files to scan',
          type: 'string'
        }),
      handler: async (args: Args): Promise<void> => {
        if (await scanFiles(args)) {
          process.exit(1);
        }
      }
    })
    .demandCommand(1)
    .option('exclude', {
      default: [],
      describe: 'exclude pattern',
      type: 'array'
    })
    .option('luhn-check', {
      default: true,
      describe: 'check if suspected card number passes the luhn check',
      type: 'boolean'
    })
    .option('silent', {
      describe: 'do not show any output',
      type: 'boolean'
    })
    .option('verbose', {
      describe: 'show all scanned files',
      type: 'boolean'
    })
    .conflicts('silent', 'verbose')
    .help().argv;
};

export default run;
export { run, scanFiles, scanFile, showMatches, isCardNumber };
