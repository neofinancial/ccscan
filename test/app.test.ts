import stripAnsi from 'strip-ansi';

import { scanFiles } from '../src/app';

let consoleLogSpy: jest.SpyInstance;

const formatConsoleOutput = (consoleOutput: string[][]): string[] => {
  return consoleOutput.map((line: string[]): string => stripAnsi(line[0]));
};

beforeAll(() => {
  consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
});

afterAll(() => {
  jest.restoreAllMocks();
});

afterEach(() => {
  jest.resetAllMocks();
});

describe('TypeScript', () => {
  test('card numbers are detected', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/typescript/*.{ts,js}' })).toEqual([
      'test/fixtures/typescript/bad.ts'
    ]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(12);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('card numbers are detected but ignored numbers are not included', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/typescript/*.{ts,js}', ignoreNumbers: ["5555554240233167"] })).toEqual([
      'test/fixtures/typescript/bad.ts'
    ]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toHaveLength(0)
  });

  test('card numbers are detected in silent mode', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/typescript/*.{ts,js}', silent: true })
    ).toEqual(['test/fixtures/typescript/bad.ts']);
    expect(consoleLogSpy).not.toHaveBeenCalled();
  });

  test('card numbers are detected in verbose mode', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/typescript/*.{ts,js}', verbose: true })
    ).toEqual(['test/fixtures/typescript/bad.ts']);
    expect(consoleLogSpy).toHaveBeenCalledTimes(14);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('card-like numbers are detected when luhn check is disabled', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/typescript/*.{ts,js}', luhnCheck : false })
    ).toEqual(['test/fixtures/typescript/bad.ts', 'test/fixtures/typescript/questionable.ts']);
    expect(consoleLogSpy).toHaveBeenCalledTimes(24);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('card numbers are not detected when no files match pattern', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/typescript/*.js' })).toEqual([]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
  });

  test('card numbers are not detected in excluded files', async () => {
    expect(
      await scanFiles({
        files: '**/test/fixtures/typescript/*.{ts,js}',
        exclude: ['test/fixtures/typescript']
      })
    ).toEqual([]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
  });
});

describe('JavaScript', () => {
  test('card numbers are detected', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/javascript/*.{ts,js}' })).toEqual([
      'test/fixtures/javascript/bad.js'
    ]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(12);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('files with ignored card numbers are not included', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/javascript/*.{ts,js}' , ignoreNumbers: ["5555554240233167"] })).toEqual([]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toHaveLength(0)
  });

  test('card numbers are detected in silent mode', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/javascript/*.{ts,js}', silent: true })
    ).toEqual(['test/fixtures/javascript/bad.js']);
    expect(consoleLogSpy).not.toHaveBeenCalled();
  });

  test('card numbers are detected in verbose mode', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/javascript/*.{ts,js}', verbose: true })
    ).toEqual(['test/fixtures/javascript/bad.js']);
    expect(consoleLogSpy).toHaveBeenCalledTimes(14);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('card-like numbers are detected when luhn check is disabled', async () => {
    expect(
      await scanFiles({ files: '**/test/fixtures/javascript/*.{ts,js}', luhnCheck: false })
    ).toEqual(['test/fixtures/javascript/bad.js', 'test/fixtures/javascript/questionable.js']);
    expect(consoleLogSpy).toHaveBeenCalledTimes(24);
    expect(formatConsoleOutput(consoleLogSpy.mock.calls)).toMatchSnapshot();
  });

  test('card numbers are not detected when no files match pattern', async () => {
    expect(await scanFiles({ files: '**/test/fixtures/javascript/*.ts' })).toEqual([]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
  });

  test('card numbers are not detected in excluded files', async () => {
    expect(
      await scanFiles({
        files: '**/test/fixtures/javascript/*.{ts,js}',
        exclude: ['test/fixtures/javascript']
      })
    ).toEqual([]);
    expect(consoleLogSpy).toHaveBeenCalledTimes(0);
  });
});
