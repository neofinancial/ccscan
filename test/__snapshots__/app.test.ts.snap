// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JavaScript card numbers are detected 1`] = `
[
  "test/fixtures/javascript/bad.js",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
]
`;

exports[`JavaScript card numbers are detected in verbose mode 1`] = `
[
  "test/fixtures/javascript/bad.js",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
  "test/fixtures/javascript/good.js",
  "test/fixtures/javascript/questionable.js",
]
`;

exports[`JavaScript card-like numbers are detected when luhn check is disabled 1`] = `
[
  "test/fixtures/javascript/bad.js",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
  "test/fixtures/javascript/questionable.js",
  undefined,
  "  3:29     warn   // card number in a comment 1234123412341234",
  "  6:31     warn     const cardNumberNoSpaces = '1234123412341234';",
  "  7:34     warn     const cardNumberSingleSpace = '1234 1234 1234 1234';",
  "  8:33     warn     const cardNumberSingleDash = '1234-1234-1234-1234';",
  "  9:37     warn     const cardNumberMultipleSpaces = '1234  1234  1234  1234';",
  "  10:37    warn     const cardNumberMultipleDashes = '1234--1234--1234--1234';",
  "  11:54    warn     const cardNumberInString = 'This is a card number: 1234123412341234';",
  "  12:43    warn     const cardNumberInStringNoSpaces = 'text1234123412341234text';",
  "  13:33    warn     const cardNumberWithThrees = '3333333333333333';",
  undefined,
]
`;

exports[`JavaScript files with both ignored and non-ignored card numbers are included 1`] = `
[
  "test/fixtures/javascript/bad.js",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  undefined,
]
`;

exports[`TypeScript card numbers are detected 1`] = `
[
  "test/fixtures/typescript/bad.ts",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
]
`;

exports[`TypeScript card numbers are detected in verbose mode 1`] = `
[
  "test/fixtures/typescript/bad.ts",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
  "test/fixtures/typescript/good.ts",
  "test/fixtures/typescript/questionable.ts",
]
`;

exports[`TypeScript card-like numbers are detected when luhn check is disabled 1`] = `
[
  "test/fixtures/typescript/bad.ts",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  "  15:37    error    const secondCardNumberNoSpaces = '5411119998052468';",
  undefined,
  "test/fixtures/typescript/questionable.ts",
  undefined,
  "  3:29     warn   // card number in a comment 1234123412341234",
  "  6:31     warn     const cardNumberNoSpaces = '1234123412341234';",
  "  7:34     warn     const cardNumberSingleSpace = '1234 1234 1234 1234';",
  "  8:33     warn     const cardNumberSingleDash = '1234-1234-1234-1234';",
  "  9:37     warn     const cardNumberMultipleSpaces = '1234  1234  1234  1234';",
  "  10:37    warn     const cardNumberMultipleDashes = '1234--1234--1234--1234';",
  "  11:54    warn     const cardNumberInString = 'This is a card number: 1234123412341234';",
  "  12:43    warn     const cardNumberInStringNoSpaces = 'text1234123412341234text';",
  "  13:33    warn     const cardNumberWithThrees = '3333333333333333';",
  undefined,
]
`;

exports[`TypeScript files with both ignored and non-ignored card numbers are included 1`] = `
[
  "test/fixtures/typescript/bad.ts",
  undefined,
  "  3:29     error  // card number in a comment 5555554240233167",
  "  4:34     error  // two card numbers in a comment 5555554240233167 1234123412341234",
  "  7:31     error    const cardNumberNoSpaces = '5555554240233167';",
  "  8:34     error    const cardNumberSingleSpace = '5555 5542 4023 3167';",
  "  9:33     error    const cardNumberSingleDash = '5555-5542-4023-3167';",
  "  10:37    error    const cardNumberMultipleSpaces = '5555  5542  4023  3167';",
  "  11:37    error    const cardNumberMultipleDashes = '5555--5542--4023--3167';",
  "  12:54    error    const cardNumberInString = 'This is a card number: 5555554240233167';",
  "  13:43    error    const cardNumberInStringNoSpaces = 'text5555554240233167text';",
  undefined,
]
`;
